File: backend/src/main/java/at/ac/tuwien/sepr/assignment/individual/entity/Horse.java
package at.ac.tuwien.sepr.assignment.individual.entity;

import at.ac.tuwien.sepr.assignment.individual.type.Sex;
import java.time.LocalDate;

/**
 * Represents a horse entity in the persistent data store.
 * This record encapsulates all relevant attributes of a horse,
 * including personal details, ownership, lineage, and optional media information.
 *
 * @param id            the unique identifier of the horse.
 * @param name          the name of the horse.
 * @param description   a brief description or additional information about the horse.
 * @param dateOfBirth   the birthdate of the horse.
 * @param sex           the biological sex of the horse.
 * @param ownerId       the unique identifier of the horse's owner.
 * @param motherId      the unique identifier of the horse's mother (may be null if unknown).
 * @param fatherId      the unique identifier of the horse's father (may be null if unknown).
 * @param image         the binary image data representing the horse (optional, may be null).
 * @param imageType     the MIME type of the image (optional, e.g., "image/png").
 */
public record Horse(
    Long id,
    String name,
    String description,
    LocalDate dateOfBirth,
    Sex sex,
    Long ownerId,
    Long motherId,
    Long fatherId,
    byte[] image,
    String imageType
) {
}

File: frontend/src/app/dto/horse.ts
import {Owner} from './owner';
import {Sex} from './sex';

export interface Horse {
  id?: number;
  name: string;
  description?: string;
  dateOfBirth: Date;
  sex: Sex;
  owner?: Owner;
  mother?: Horse;
  father?: Horse;
  image?: File;
  imageType?: string;
}

export interface HorseCreate {
  name: string;
  description?: string;
  dateOfBirth: Date;
  sex: Sex;
  ownerId?: number;
  motherId?: number;
  fatherId?: number;
  image?: File;
  imageType?: string;
}

export function convertFromHorseToCreate(horse: Horse): HorseCreate {
  return {
    name: horse.name,
    description: horse.description,
    dateOfBirth: horse.dateOfBirth,
    sex: horse.sex,
    ownerId: horse.owner?.id,
    motherId: horse.mother?.id,
    fatherId: horse.father?.id,
  };
}

export interface HorseUpdate {
  id: number;
  name: string;
  description?: string;
  dateOfBirth: Date;
  sex: Sex;
  ownerId?: number;
  motherId?: number;
  fatherId?: number;
  image?: File;
  imageType?: string;
}

export interface HorsePedigree {
  id: number;
  name: string;
  dateOfBirth: Date;
  sex: Sex;
  mother?: HorsePedigree;
  father?: HorsePedigree;
  expanded?: boolean;
}

export function convertFromHorseToUpdate(horse: Horse): HorseUpdate {
  return {
    id: horse.id!,
    name: horse.name,
    description: horse.description,
    dateOfBirth: horse.dateOfBirth,
    sex: horse.sex,
    ownerId: horse.owner ? horse.owner.id : undefined,
    motherId: horse.mother ? horse.mother.id : undefined,
    fatherId: horse.father ? horse.father.id : undefined,
  };
}
